openupgradelib/openupgrade.py

```diff
@@ -1785,3 +1788,91 @@ def format_message(f):
                     )
                     f.active = False
                     break
+
+
+def add_fields(env, field_spec):
+    """This method adds all the needed stuff for having a new field populated in
+    the DB (SQL column, ir.model.fields entry, ir.model.data entry...).
+
+    It's intended for being run in pre-migration scripts for pre-populating
+    fields that are going to be declared later in the module.
+
+    NOTE: This only works in >=v8 and is not needed in >=v12, as now Odoo always
+    add the XML-ID entry:
+    https://github.com/odoo/odoo/blob/9201f92a4f29a53a014b462469f27b32dca8fc5a/
+    odoo/addons/base/models/ir_model.py#L794-L802
+
+    :param: field_spec: List of tuples with the following expected elements
+      for each tuple:
+
+      * field name
+      * model name
+      * SQL table name: Put `False` if the model is already loaded in the
+        registry and thus the SQL table name can be obtained that way.
+      * field type: binary, boolean, char, date, datetime, float, html, integer,
+        many2many, many2one, monetary, one2many, reference, selection, text,
+        serialized. The list can vary depending on Odoo version or custom added
+        field types.
+      * SQL field type: If the field type is custom or it's one of the special
+        cases (see below), you need to indicate here the SQL type to use
+        (from the valid PostgreSQL types):
+        https://www.postgresql.org/docs/9.6/static/datatype.html
+      * module name: for adding the XML-ID entry.
+    """
+    sql_type_mapping = {
+        'binary': 'bytea',  # If there's attachment, no SQL. Force it manually
+        'boolean': 'bool',
+        'char': 'varchar',  # Force it manually if there's size limit
+        'date': 'date',
+        'datetime': 'timestamp',
+        'float': 'numeric',  # Force manually to double precision if no digits
+        'html': 'text',
+        'integer': 'int4',
+        'many2many': False,  # No need to create SQL column
+        'many2one': 'int4',
+        'monetary': 'numeric',
+        'one2many': False,  # No need to create SQL column
+        'reference': 'varchar',
+        'selection': 'varchar',  # Can be sometimes integer. Force it manually
+        'text': 'text',
+        'serialized': 'text',
+    }
```

Comment:
can't you call https://github.com/OCA/OCB/blob/8.0/openerp/models.py#L170 here?

Reply:
That methods is not available anymore in newer versions. It has been replaced by a property on each field class, which sometimes it's a direct value (example: https://github.com/odoo/odoo/blob/9961bbf4eb87f281e8631d5471bdcad06eb35440/odoo/fields.py#L1120) and sometimes a method that requires a proper instance (example: https://github.com/odoo/odoo/blob/9961bbf4eb87f281e8631d5471bdcad06eb35440/odoo/fields.py#L1820), so it can't be easily reproducible. That's why I opted for this hardcoded option, being limited its scope in the future (it's not needed for >=v12).

Reply:
makes sense, thanks for the links